### PR TITLE
[Profiler] Add Pattern that detects extra cuda copy

### PR DESCRIPTION
**Pull request by davidchencsl**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81056
* __->__ #80572
* #80108

### CODE REVIEW

File: torch/profiler/_pattern_matcher.py [hunk 1]
```diff
@@ -69,6 +69,31 @@ def match(self, event):
...
+        event = event.parent
+        if event is None:
+            return False
+        # Check if we have a aten::fill_ in previous leaf
+        event = self.prev_of(event)
+        if event is None:
+            return False
+        while event.children:
+            event = event.children[-1]
+            if event.name() == "aten::fill_" or event.name() == "aten::zero_":
```
Comment (robieta) [inline torch/profiler/_pattern_matcher.py:92]: I thought `zero_` called into `fill_` to actually implement the compute?

File: test/test_profiler.py [hunk 2]
```diff
@@ -1317,9 +1317,38 @@ def test_profiler_pattern_match_helper(self):
...
         self.assertEqual(event_tree[0], pattern.prev_of(event_tree[1]))
 
-
-
-
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA is required")
+    def test_profiler_extra_cuda_copy_pattern(self):
+        from torch.profiler._pattern_matcher import ExtraCUDACopyPattern
+
+        with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
```
Comment (robieta) [inline test/test_profiler.py:1324]: nit: do we need `activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA]` on all of them? (It's just cluttering the screen a bit)

**Review by robieta — CHANGES_REQUESTED** [review]
I think the traversal needs a bit of work, but overall on the right track.

File: torch/profiler/_pattern_matcher.py (hunk 1 above)
Comment (davidchencsl) [reply torch/profiler/_pattern_matcher.py:92]: It turn out that zero_ is a special case and it dosen't call fill_ :(

**Review by robieta — APPROVED** [review]
Looks great.

**Pull request closed without merging**